File: live_coding/deno_core_live/src/ops/fetch.js
((window) => {
  async function fetch(args) {
    if (typeof args === "string") {
      args = { url: args, method: "GET", headers: [], body: null };
    } else if (typeof args === "object") {
      if (args.url) {
        args.method = args.method || "GET";
        args.headers = args.headers || [];
        args.body = args.body || null;
      } else {
        throw new Error("Invalid arguments");
      }
    } else {
      throw new Error("Invalid fetch args, should be string or object");
    }
    let res = await Deno.core.opAsync("op_fetch", args);
    res.text = () => {
      let body = res.body;
      if (!body) return null;
      return Deno.core.opSync("op_decode_utf8", body);
    };
    res.json = () => {
      let text = res.text();
      if (!text) return null;
      return JSON.parse(text);
    };
    return res;
  }

  window.fetch = fetch;
})(this);

File: live_coding/deno_core_live/src/ops/fetch.rs
use deno_core::{
    error::AnyError, include_js_files, op, ByteString, Extension, OpState, ZeroCopyBuf,
};
use reqwest::{
    header::{HeaderName, HeaderValue},
    Method, Url,
};
use serde::{Deserialize, Serialize};
use std::{cell::RefCell, rc::Rc, str::FromStr};

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FetchArgs {
    method: String,
    url: String,
    headers: Vec<(ByteString, ByteString)>,
    body: Option<ZeroCopyBuf>,
}

#[derive(Serialize)]
pub struct FetchResponse {
    status: u16,
    status_text: String,
    headers: Vec<(ByteString, ByteString)>,
    body: Option<ZeroCopyBuf>,
}

#[op]
async fn op_fetch(state: Rc<RefCell<OpState>>, args: FetchArgs) -> Result<FetchResponse, AnyError> {
    let state_ref = state.borrow();
    let client = state_ref.borrow::<reqwest::Client>().clone();
    let method = Method::from_str(&args.method.to_ascii_uppercase())?;
    let url = Url::parse(&args.url)?;
    let mut req = client.request(method, url);
    for (k, v) in args.headers {
        req = req.header(HeaderName::from_bytes(&k)?, HeaderValue::from_bytes(&v)?);
    }

    let req = if let Some(body) = args.body {
        req.body(Vec::from(&*body))
    } else {
        req
    };
    let res = req.send().await?;

    let status = res.status().as_u16();
    let status_text = res.status().canonical_reason().unwrap_or("").to_string();
    let headers = res
        .headers()
        .iter()
        .map(|(k, v)| (k.as_str().into(), v.as_bytes().into()))
        .collect();

    let body = res.bytes().await?;
    let body = if body.is_empty() {
        None
    } else {
        Some(body.to_vec().into())
    };

    Ok(FetchResponse {
        status,
        status_text,
        headers,
        body,
    })
}

pub fn init() -> Extension {
    Extension::builder()
        .js(include_js_files!(
            prefix "fetch",
            "src/ops/fetch.js",
        ))
        .ops(vec![op_fetch::decl()])
        .state(move |state| {
            state.put::<reqwest::Client>(reqwest::Client::new());
            Ok(())
        })
        .build()
}
